### ISSUE TITLE
sf_auth(login_url = ) Bad Request (HTTP 400)

### ISSUE THREAD

**Issue by rdinter-usda** [StevenMMortimer/salesforcer#132]
Issues with authentication for private instance of Salesforce that appears similar to https://github.com/StevenMMortimer/salesforcer/issues/111 and https://github.com/StevenMMortimer/salesforcer/issues/108.

```r
> sf_auth(login_url = "https://PRIVATE-INSTANCE.lightning.force.com/")
Waiting for authentication in browser...
Press Esc/Ctrl + C to abort
Authentication complete.
Error in oauth2.0_access_token(endpoint, app, code = code, user_params = user_params,  : 
  Bad Request (HTTP 400). Failed to get an access token.
```

Authentication appears to work through the browser, however this error shows up in R. What is strange is I can see a file `.httr-oauth-salesforcer` in working directory that is created and it is roughly 6 bytes. Then when I attempt to try `sf_list_reports()` another window pops up for authentication, the browser says authentication worked, and once I return to R the report list is displayed.

**Comment by StevenMMortimer** [comment]
Is this still an issue @rdinter-usda? Based on what you described, it prompts twice and works the second time?

I will also be switching the package over from {{httr}} to the {{httr2}} package, which maybe could help some of these transient errors if it handles the authentication in a more sophisticated way. I'm not sure.

**Comment by rdinter-usda** [comment]
Yes this is still an issue, although I would not say it works the second time. Each time I call `sf_auth(login_url = )` it fails with an error. However, if I call other functions used in salesforceR like `sf_list_reports()` then it succeeds. I believe that's related to some side-effect of cacheing of the oauth.

**Comment by StevenMMortimer** [comment]
Are you accessing your R session on a remote server via the browser? In other words, are you not running it locally on your laptop? Or is it possible that some other process is running on port 1410 of your machine?

According to [this article](https://cran.r-project.org/web/packages/gargle/vignettes/auth-from-web.html), it sounds like either of those would cause issues like you're seeing and you'd have to try and use out of band authentication. I can try to add those arguments to the `sf_auth()` function so they can be passed through to {{httr}} package functions.

**Comment by dethmasque** [comment]
I was seeing the same behavior when running `salesforcer::sf_auth(login_url="https://example.my.salesforce.com")` locally. The error in the URL implied that this is PKCE-related: `http://localhost:1410/?error=invalid_request&error_description=missing+required+code+challenge`. Turning off PKCE at the organization level resolved the error and allowed for successful authentication:

![Screenshot 2024-12-09 at 4 00 59 PM](https://github.com/user-attachments/assets/386aeabd-19bf-4528-8249-66b4c64a3530)

![image](https://github.com/user-attachments/assets/9dfeadfc-8c48-4cc5-b517-816f49ef1fea)

```
 salesforcer::sf_auth(login_url = "https://example.my.salesforce.com")
Waiting for authentication in browser...
Press Esc/Ctrl + C to abort
Authentication complete.
```

**Comment by StevenMMortimer** [comment]
@dethmasque Really appreciate your comment about PKCE! I'll add that to the function documentation and getting started article. Thanks again!

cc @rdinter-usda
